### PR TITLE
Fix some core typeof(...) references

### LINKED ISSUE #105772: DefineInitializedData on TypeBuilder in System.Reflection.Emit uses typeof(ValueType) internally even when using a MetadataLoadContext
### Description

DefineInitializedData on TypeBuilder in System.Reflection.Emit uses typeof(ValueType) internally, even when using a MetadataLoadContext.

This causes type.IsValueType to return false even though it is true, which causes invalid IL to be emitted (fields are being marked `class` when they should be marked `valuetype`, causing failures to load)

### Reproduction Steps

```csharp
using System.Reflection;
using System.Reflection.Emit;

var pathres = new PathAssemblyResolver([typeof(object).Assembly.Location]);
var mlc = new MetadataLoadContext(pathres, typeof(object).Assembly.GetName().Name);

PersistedAssemblyBuilder pab = new PersistedAssemblyBuilder(
    new AssemblyName("MyAssembly"),
    mlc.CoreAssembly!
);

ModuleBuilder mb = pab.DefineDynamicModule("Module");
TypeBuilder tb = mb.DefineType("Type");

var field = tb.DefineUninitializedData("UninitializedDataField", 100, FieldAttributes.Public);
var type = field.FieldType.BaseType;

Console.WriteLine(type.IsValueType);
Console.WriteLine(type);
Console.WriteLine(type == typeof(ValueType));

var contextValueType = mlc.CoreAssembly!.GetType("ValueType")!;
Console.WriteLine(type == contextValueType);
```
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Reflection.MetadataLoadContext" Version="8.0.0" />
  </ItemGroup>

</Project>
```

### Expected behavior

type.IsValueType should return true.
type == typeof(ValueType) should return false.
type == contextValueType should return true.

### Actual behavior

type.IsValueType returns false.
type == typeof(ValueType) returns true.
type == contextValueType returns false.

### Regression?

Unknown

### Known Workarounds

_No response_

### Configuration

Arch Linux x64, dotnet 9.0.100-rc.1.24381.3

### Other information

As far as I can tell, this appears to be because DefineDataHelper in TypeBuilderImpl.cs uses typeof(ValueType) to define its helper type.

### PR DESCRIPTION
**Pull request by buyaa-n**
- Fix `typeof(ValueType)` usage in `DefineUninitializedData` on `TypeBuilder` mentioned in https://github.com/dotnet/runtime/issues/105772
- Fix another similar case in `GetArrayMethod` on `ModuleBuidler` that uses `typeof(void)`
- There is another case for creating `EnumBuilderImpl` constructor where `typeof(Enum)` is used as parent type, but according to the tests I added it appears to be fine, no need to change `typeof(Enum)` reference here https://github.com/dotnet/runtime/blob/f120fff4838fbae622123c6578734275d909a3f2/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/EnumBuilderImpl.cs#L14-L20

Fixes https://github.com/dotnet/runtime/issues/105772

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs [hunk 1]
```diff
@@ -439,6 +439,7 @@ protected override FieldBuilder DefineUninitializedDataCore(string name, int siz
             return DefineDataHelper(name, new byte[size], size, attributes);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:DynamicallyAccessedMembers", Justification = "Types are preserved via ModuleBuilderImpl.s_coreTypes")]
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:442]: This warning is complaining about members not being preserved. `ModuleBuilderImpl.s_coreTypes` does not preserve the members, so this justification is not quite right.

Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:442]: I do not think there is anything real to fix, except the text of the justification.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:442]: No any member of `ValueType` is referenced in the method, how about updating the message to `ValueType members are not referenced in this context`?

Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:442]: LGTM

**Review by steveharter — APPROVED** [review]

**Comment by steveharter** [comment]
Buyaa out, merging.

**Pull request merged**